File: idletime.pl
#!/usr/bin/perl
#	 ___    _ _       _____ _               
#	|_ _|__| | |___  |_   _(_)_ __  ___ _ _ 
#	 | |/ _` | / -_)   | | | | '  \/ -_| '_|
#	|___\__,_|_\___|   |_| |_|_|_|_\___|_|  
#	v1.1 [2016-01-22]
#	by Jack the Smack                                      
#	Old Discussion Thread: http://forum.teamspeak.com/showthread.php/101272-Perl-Idle-Timer-Display-how-long-they-ve-been-AFK-without-moving-or-kicking-them
#	Github: https://github.com/jackthesmack/IdleTime
#	Contact: idle_timer@jack-cole.com

#	How to use: 
#		Configure all the settings in idletime.cfg, then run this:
#			perl "...\idletime.pl"
#		in terminal, or download Perl for Windows by using your google foo. 

##		----------------------------		##
##		Ignore everything below here		##
##		...or not.							##
##		I'm a comment, not your mother.		##
##		----------------------------		##

use strict; use warnings;
use Data::Dumper qw(Dumper);
use File::Basename;
use Cwd 'abs_path';
use Time::HiRes qw (usleep);
no warnings "experimental::autoderef";

#Defaults
my %cfg = (
	"host" => 'localhost',
	"port" => '10011',
	"username" => "serverQuery",
	"password" => "xxx",
	"sid" => 1,
	"minidleminutes" => 25,
	"labelhours" => " Hr",
	"labelminutes" => " Min",
	"labelseconds" => " Sec",
	"labelplural" => 1,
	"idletitle" => " Idle",
	"idlenickname" => "Idle Timer",
	"shortenedtime" => 0,
	"secondsenabled" => 0,
	"serversortid" => 91115,
	"updatetime" => 5,
	"testing" => 0,
	"slowdown" => 0
);


#CFG File
my $cfgfilename = dirname(abs_path($0))."/";
$cfgfilename .= 'idletime.cfg';
if (open(my $fh, $cfgfilename)) {
  while (my $row = <$fh>) {
	chomp $row;
	$row =~ s/[^\\]\#[^\n]+//g;
	if($row =~ /[^\n]+=[^\n]+/){
		my @row = split(/\=/, $row);
		next if(@row != 2);
		$row[0] = trim($row[0]);$row[1] = trim($row[1]);
		$row[0] = lc($row[0]);
		$row[1] = ($row[1] =~ /[\"\']([^\n]+)[\"\']/)[0] if($row[1] =~ /[\"\'][^\n]+[\"\']/);
		$cfg{$row[0]} = $row[1];
		
	};
  };
}else{
	warn "Could not open file '$cfgfilename' $!\n";
};


#Command Line Arguments
foreach (@ARGV){
	my ($Key) = $_ =~ /([^ ]+)=[^ ]+/;
	my ($Value) = $_ =~ /[^ ]+=([^\n]+)/;
	$Key = lc($Key);
	$cfg{$Key} = $Value;
};


$cfg{"idlenickname"}  =~ s/ /\\s/g;

print "Connecting to server\n";
my $buf = '';
use IO::Socket; 
my $sock = new IO::Socket::INET (
	PeerAddr => $cfg{"host"}
	,PeerPort => $cfg{"port"}
	,Proto => 'tcp'
	,Autoflush   => 1
	,Blocking    => 1
	,Timeout    => 10
	); 
die "Could not connect to Server: $!\n" unless $sock;



sub ExecuteCommand{ # ExecuteCommand( command [string], return error [boolean] )
	$_[1] = 0 if(!defined($_[1]));
	print $sock $_[0]."\n";
	my $response = "";
	while (1){
		$sock->sysread($buf,1024*10);
		$response .= $buf;
		if($buf =~ /error id=\d+/){
			if(!$_[1] == 1){
				my @split_response = split("error id", $response);
				$response = $split_response[0];
			};
			last;
		};
	};
	usleeep_(1);
	return $response;

};

sub ConvertTime{
	my $IdleSeconds = floor($_[0]/1000);
	my $IdleMinutes =  floor($IdleSeconds/60);
	my $IdleHours =  floor($IdleMinutes/60);
	$IdleSeconds = $IdleSeconds - $IdleMinutes*60;
	$IdleMinutes = $IdleMinutes - $IdleHours*60;
	my @returnTime;
	push(@returnTime, "$IdleHours$cfg{labelhours}".Pluralize($IdleHours)) if($IdleHours > 0 );
	push(@returnTime, "$IdleMinutes$cfg{labelminutes}".Pluralize($IdleMinutes)) if($IdleMinutes > 0);
	push(@returnTime, "$IdleSeconds$cfg{labelseconds}".Pluralize($IdleSeconds)) if($_[1] == 1);
	return join(' ', @returnTime) if($_[2] == 0);
	return $returnTime[0] if($_[2] == 1);
	

};
sub ParseError{
	my ($error_msg) = $_[0] =~ /msg=([^\t]+)/;
	$error_msg =~ s/ extra_msg=/ | /;
	$error_msg =~ s/\\s/ /g;
	$error_msg =~ s/\n\s/ /g;
	return $error_msg;
}
sub Pluralize{	
	return "" if ($_[0] == 1 || !$cfg{"labelplural"});
	return "s";
};
sub floor{
	return ($_[0] =~ /(\d+)/i)[0];
};

sub usleeep_{
	usleep($cfg{"slowdown"}*$_[0]) if($cfg{"slowdown"} > 0);
};
sub  trim { my $s = shift; $s =~ s/^\s+|\s+$//g; return $s };
my $LoginResponse = ExecuteCommand("login $cfg{username} $cfg{password}",1);

my $TS3_Confirmed = $LoginResponse =~ /TS3\n/;
my $Login_Confirmed = $LoginResponse =~ /error id=0/;
my $Login_error = ParseError($LoginResponse);


print "Successfully connected to Teamspeak 3 Server\n" if ($TS3_Confirmed);
die "Error: Did not connect to a Teamspeak 3 server\n" if (!$TS3_Confirmed);
print "Login Successful\n" if ($Login_Confirmed);
die "Error: $Login_error\n" if(!$Login_Confirmed);

my $SID_Response = ExecuteCommand("use sid=$cfg{sid}", 1);
my $SID_Confirmed = $SID_Response =~ /error id=0/;
my $SID_error = ParseError($SID_Response);
print "Connected to SID $cfg{sid}\n" if ($SID_Confirmed);
die "SID Error: $SID_error\n" if(!$SID_Confirmed);

print ExecuteCommand("clientupdate client_nickname=$cfg{idlenickname}")."\n";



while($sock){
	print "\[".localtime."\] Beginning Loop...\n";
	my $client_list = ExecuteCommand("clientlist",1);
	if(!$client_list =~ /error code=0/i){ #Skips if no client information
		print "\[".localtime."\] No Client Information retrieved";
		print "--- Waiting $cfg{updatetime} second".Pluralize($cfg{"updatetime"})." before next execution\n" if ($cfg{"testing"});
		sleep($cfg{"updatetime"});
	};
	my @client_IDs = $client_list =~ /clid=(\d+)/g;	
	my $idle_times;
	my $ClientIdleCount = 0;
	my %IdleServerGroups;
	my %Clients;
	

	foreach (@client_IDs) {
		my $client_info = ExecuteCommand("clientinfo clid=$_",1);
		print "$client_info \n" if($cfg{'testing'});
		next if($client_info =~ /client_platform=ServerQuery/i); #Skips if server Query
		my ($client_idle) = $client_info =~ /client_idle_time=(\d+)/i;
		my($client_name) = $client_info =~ /client_nickname=([a-zA-Z0-9\\]{1,})/i;
		my ($client_db_id) = $client_info =~ /client_database_id=(\d+)/i;
		$client_name =~ s/\\s/ /g;
		$idle_times .=  "\t[$_]\t[$client_db_id]\t\t$client_name\t$client_idle ms\n";
		$Clients{"$client_db_id"}{"client_id"} = $_;
		$Clients{"$client_db_id"}{"client_db_id"} = $client_db_id;
		$Clients{"$client_db_id"}{"client_name"} = $client_name;
		$Clients{"$client_db_id"}{"client_idle"} = $client_idle;
		
	}; 	

	my @servergrouplist = split(/\|/, ExecuteCommand("servergrouplist",1));

	foreach (@servergrouplist) {
		
		my $servergroupid = ($_ =~ /sgid=(\d+)/)[0];
		if(length($servergroupid) == 0){
			next;
		}
		my $server_group_sort_group_id = ($_ =~ /sortid=(\d+)/)[0];
		next if(!($server_group_sort_group_id == $cfg{"serversortid"}));
		my $sgclientlist = ExecuteCommand("servergroupclientlist sgid=$servergroupid");
		# print "\$\"servergroupclientlist sgid=$servergroupid\": ".$sgclientlist."\n";
		# print "Press Enter to continue\n";
		# <STDIN>;
		if (!length($sgclientlist =~ s/^\s+//) == 0){
			print "Deleting Empty Group: ".ExecuteCommand("servergroupdel sgid=$servergroupid force=1")."\n";
			next;
		};
		# my @sg_client_db_id = split(/\|/, $sgclientlist);
		# $IdleServerGroups{"$servergroupid"}{"clients"} = "$sgclientlist";
		$IdleServerGroups{"$servergroupid"}{"name"} = ($_ =~ /name=([a-zA-Z\\0-9]{1,})/)[0];
		$IdleServerGroups{"$servergroupid"}{"response"} = ($_);
		
		my @sg_Client_list = split(/\|/, $sgclientlist);

		foreach my $Client_db_id_key (@sg_Client_list) {
			($Client_db_id_key)  = $Client_db_id_key =~ /cldbid=(\d+)/;
			$Clients{"$Client_db_id_key"}{"server_idle_group"} = $servergroupid;
		};
		if(!@sg_Client_list){
			print "(2)Deleting Empty Group sgid=$servergroupid: ".ExecuteCommand("servergroupdel sgid=$servergroupid force=1")."\n";
		};
		
		
	};
	print 	"\$Clients: \n\t".(Dumper \%Clients)."\n" if ($cfg{"testing"});
	print 	"\$IdleServerGroups: \n\t".(Dumper \%IdleServerGroups)."\n" if ($cfg{"testing"});
	
	my $total_clients = 0;
	my $client_verbose_info = "" if ($cfg{"testing"});
	foreach my $Client_key (keys {%Clients}) {
		
		
		my $client_id = $Clients{$Client_key}{"client_id"};
		my $client_db_id = $Clients{$Client_key}{"client_db_id"};
		my $client_name = $Clients{$Client_key}{"client_name"};
		$client_name =~ s/\\s/ /g;
		my $client_idle = $Clients{$Client_key}{"client_idle"};
		my $server_idle_group = "";
		$server_idle_group = $Clients{$Client_key}{"server_idle_group"} if(exists($Clients{$Client_key}{"server_idle_group"}));
		
		
		# if($client_db_id == 4 && $client_idle > 1000*5){  #For testing on a single client
		if ($client_idle > 1000*60*$cfg{"minidleminutes"}){
			# Client is Idle
			$ClientIdleCount++;
			my $client_idle_time = ConvertTime($client_idle,$cfg{"secondsenabled"},$cfg{"shortenedtime"});
			my $client_idle_time_formatted = $client_idle_time.$cfg{"idletitle"};
			$client_idle_time_formatted =~ s/ /\\s/g;
			if(exists($Clients{$Client_key}{"server_idle_group"})){
				my $server_idle_group_name = $IdleServerGroups{"$server_idle_group"}{"name"};
				if($server_idle_group_name eq $client_idle_time_formatted){
					print "Server Group $server_idle_group\'s name \"$client_idle_time\" doesn't need to be changed\n" if ($cfg{"testing"});
				}else{
					my $server_idle_rename = ExecuteCommand("servergrouprename sgid=$server_idle_group name=$client_idle_time_formatted",1);
					if($server_idle_rename =~ /error id=1282/){
						ExecuteCommand("servergroupdel sgid=$server_idle_group force=1");
						print "Server Group $server_idle_group deleted due to being unable to be renamed\n" if ($cfg{"testing"});
					}else{
						print "Server Group $server_idle_group renamed to \"$client_idle_time\"\n" if ($cfg{"testing"});
					};
				};
			}else{
				my $server_idle_add = ExecuteCommand("servergroupadd name=$client_idle_time_formatted", 1);
				if($server_idle_add =~ /error id=1282/){
					foreach my $IdleServerGroupKey (keys {%IdleServerGroups}) {
						$server_idle_group = $IdleServerGroupKey if($IdleServerGroups{$IdleServerGroupKey}{"name"} eq $client_idle_time_formatted);
					};
				}else{
					($server_idle_group) = $server_idle_add =~ /sgid=(\d+)/;
					$IdleServerGroups{$server_idle_group}{"name"} = $client_idle_time_formatted;
				};
				ExecuteCommand("servergroupaddperm sgid=$server_idle_group permsid=i_group_sort_id permvalue=$cfg{serversortid} permnegated=0 permskip=0|permsid=i_group_show_name_in_tree permvalue=2 permnegated=0 permskip=0");
				print "$client_name is now marked as idle after being idle for ".ConvertTime($client_idle,1,0)."\n";
			};
			ExecuteCommand("servergroupaddclient sgid=$server_idle_group cldbid=$client_db_id");
		}else{
			# Client is not idle
			if(exists($Clients{$Client_key}{"server_idle_group"})){
				ExecuteCommand("servergroupdelclient sgid=$server_idle_group cldbid=$Client_key");
				undef $Clients{$Client_key}{"server_idle_group"};
				print "$client_name is no longer idle.\n";
			};
		};
		
		$total_clients++;
		$client_verbose_info .= "\t$client_name\t".ConvertTime($client_idle,1,0)."\n" if ($cfg{"testing"});
		
	};
	print "\[".localtime."\] $total_clients Total Clients | ";
	print "$ClientIdleCount Idle for longer than $cfg{minidleminutes}$cfg{labelminutes}".Pluralize($cfg{"minidleminutes"})."\t\n";
	print "\t[name]\t[Current Idle Time]\n$client_verbose_info" if ($cfg{"testing"});
	print "--- Waiting $cfg{updatetime} second".Pluralize($cfg{"updatetime"})." before next execution\n" if ($cfg{"testing"});
	sleep($cfg{"updatetime"});
	
};

close $sock;
die "Connection Lost\n";

File: idleTime.py
import teamspeak3
import time
import re

hostname	= '192.168.1.10'
portnum		= '10011'
username	= 'IdleTime'
password	= 'nUpFfgav'
serverID	= '1'
nickname	= 'IdleTime'
reconnect   = True # if True it will try to reconnect if the connection is lost. False will terminate the program after connection lost.
servergroupSortID = '91115'
minIdleMinutes = 25

ts3 = teamspeak3.TeamSpeak3()

recon = 1 # this is set to reconnect's value after the first loop starts
# main program loop. Continously run until exited


def checkIdleClients(clientsList, serverGroupList):
	'''

	:type clientsList: Teamspeak3IdleClient[]
	:type serverGroupList: Teamspeak3IdleServerGroup[]
	'''
	for client in clientsList:

		# ignore self
		if( client.name == username):
			continue

		# client is not idle
		if(client.idle < minIdleMinutes * 60 * 1000):

			# client is in idle group
			for serverGroupID in client.returnServerGroupsClientIsIdleIn(serverGroupList):
				ts3.executeCommand('servergroupdelclient sgid='+serverGroupID.id+' cldbid='+client.databaseid)
				serverGroupID.clients.remove(client.databaseid)

		# client is idle
		else:

			timeIdleObj = IdleTime(client.idle)
			clientsServerGroups = client.returnServerGroupsClientIsIdleIn(serverGroupList)
			# client is in idle group
			if(len(clientsServerGroups) > 0):

				# rename the server group he is in if it needs to be renamed. Do not reduce the time.
				idlename = timeIdleObj.toString()
				if(timeIdleObj.compareToString(clientsServerGroups[0].name) > 0):
					ts3.executeCommand('servergrouprename sgid='+clientsServerGroups[0].id+' name=' + ts3.escapeString(idlename))
					clientsServerGroups[0].name = idlename


			#client is not in idle group
			else:

				# create new group and add client to it
				ts3.executeCommand('servergroupadd name=' + ts3.escapeString(timeIdleObj.toString()))
				errorCode = ts3.getErrorCode()

				# Server group add was successful
				if (errorCode == '0'):
					serverGroupID = ts3.parseLastMsg()[0].get('sgid')

				# if same server group name exists, search through the groups to find the same named one
				elif(errorCode == '1282'):
					for serverGroupToAddTo in serverGroupList:
						if(timeIdleObj.toString() == serverGroupToAddTo.name):
							serverGroupID = serverGroupToAddTo.id



				ts3.executeCommand('servergroupaddperm sgid='+serverGroupID+' permsid=i_group_sort_id permvalue='+servergroupSortID+' permnegated=0 permskip=0|permsid=i_group_show_name_in_tree permvalue=2 permnegated=0 permskip=0')
				ts3.executeCommand('servergroupaddclient sgid='+serverGroupID+' cldbid=' + client.databaseid)
	return


def removeOfflineClients(clients, idleServerGroups):
	'''

	:type clientsList: list[Teamspeak3IdleClient]
	:type serverGroupList: list[Teamspeak3IdleServerGroup]
	'''


	for serverGroup in idleServerGroups:
		for sgClient in serverGroup.clients:
			clientIsOnline = False
			for onlineClient in  clients:
				if(onlineClient.databaseid == sgClient and onlineClient.name != nickname):
					clientIsOnline = True
			if(clientIsOnline == False):
				ts3.executeCommand('servergroupdelclient sgid='+serverGroup.id+' cldbid='+sgClient)



class IdleTime:
	seconds = 0
	minutes = 0
	hours = 0

	def __init__(self, ms):
		ms = int(ms)

		self.seconds = int(ms / 1000)
		self.minutes = int(self.seconds / 60)
		self.hours = int(self.minutes / 60)
		self.seconds = self.seconds % 60
		self.minutes = self.minutes % 60
		self.hours = self.hours % 60

	def toString(self):
		returnTime = []
		if(self.hours > 0):
			returnTime.append(str(self.hours) + ' Hr' + ('' if self.hours == 1 else 's'))
		returnTime.append(str(self.minutes) + ' Min' + ('' if self.minutes == 1 else 's'))


		return ' '.join(returnTime) + ' Idle'

	def compareMinAccuracy(self, otherIdleTime):
		'''
		Compares to another IdleTime object

		:param otherIdleTime: IdleTime object to compare to.
		:type otherIdleTime: IdleTime
		:return: Positive if self is greater, zero if equal, and negative if self is less than.
		'''

		return (self.minutes - otherIdleTime.minutes) + ( 60 * (self.hours - otherIdleTime.hours) )

	def compareToString(self, otherIdleString):
		'''
		Compares to another IdleTime object that is in string format

		:param otherIdleString: Idletime that was converted toString()
		:type otherIdleString: string
		:return: Positive if self is greater, zero if equal, and negative if self is less than.
		'''
		hrSearch = re.search('(\d+) Hr', otherIdleString)
		if(hrSearch != None):
			otherHours = int(hrSearch.group(1))
		else:
			otherHours = 0

		minSearch = re.search('(\d+) Min', otherIdleString)
		if (minSearch != None):
			otherMinutes = int(minSearch.group(1))
		else:
			otherMinutes = 0

		return (self.minutes - otherMinutes) + (60 * (self.hours - otherHours))

while(recon):
	recon = reconnect
	ts3.verbose(False)

	try:
		print("Connecting to server.")
		ts3.connect(hostname, portnum)
		ts3.login(username, password, serverID, nickname)
		print("Connection successful! IdleTime is now running...")


		while(ts3.connected()):
			try:
				# Grab list of clients
				clients = ts3.getClientIdleList()

				# Grab list of idle server groups and delete empty ones
				idleServerGroups = ts3.getServerGroupbySortID(servergroupSortID)

				# Check each client to see if they're idle, then add them to a sever group if they are, or rename a server group they are already in. If they aren't idle, remove them from the group.
				checkIdleClients(clients, idleServerGroups)

				# Remove offine clients from server groups
				removeOfflineClients(clients, idleServerGroups)

				time.sleep(2)
			except ValueError as err:
				print("ValueError," , err.args)



	except ValueError as err:
		print("ValueError," , err.args)
		ts3.disconnect()
	except:
		print('Unidentified error. Disconnecting.')
		ts3.disconnect()

	if(recon):
		print('Reconnecting in 30 seconds')
		time.sleep(30)

ts3.disconnect()